kotlinx.coroutines.flow.internal.ChildCancelledException: Child of the scoped flow was cancelled

Hi, in my app I randomly get a `ChildCancellationException`. It comes from the `flatMapLatest` operator (previously named `switchMap`). I tested it on the JVM with versions 1.3.0-M2, 1.3.0-RC, 1.3.0-RC2. I was able to reproduce it in a test:
```kotlin
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.ConflatedBroadcastChannel
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlin.test.Test

class MyTest {

    @Test
    fun test(): Unit = runBlocking {
        val context = Dispatchers.Default

        val deferred = GlobalScope.async(context) {
            launch(context) {
                observeFlow().collect {}
            }

            while (true) {
                channel.send(channel.value)
            }
        }

        deferred.await()
    }

    val channel = ConflatedBroadcastChannel("")
    val flow = channel.asFlow()

    fun observeFlow() = flow.flatMapLatest {
        val flows = List(2) {
            flow {
                while (true) {
                    emit("")
                }
            }
        }
        combine(flows) { it.asList() }
    }
}
```
```
Exception in thread "DefaultDispatcher-worker-5 @coroutine#716" kotlinx.coroutines.flow.internal.ChildCancelledException: Child of the scoped flow was cancelled
	(Coroutine boundary)
	at kotlinx.coroutines.channels.AbstractChannel.registerSelectReceiveOrNull(AbstractChannel.kt:753)
	at kotlinx.coroutines.channels.AbstractChannel.access$registerSelectReceiveOrNull(AbstractChannel.kt:484)
	at kotlinx.coroutines.channels.AbstractChannel$onReceiveOrNull$1.registerSelectClause1(AbstractChannel.kt:732)
	at kotlinx.coroutines.selects.SelectBuilderImpl.invoke(Select.kt:415)
	at kotlinx.coroutines.flow.internal.CombineKt$combineInternal$2.invokeSuspend(Combine.kt:151)
Caused by: kotlinx.coroutines.flow.internal.ChildCancelledException: Child of the scoped flow was cancelled
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3$invokeSuspend$$inlined$collect$1.emit(Collect.kt:137)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.emitAll(Channels.kt:56)
	at kotlinx.coroutines.flow.FlowKt.emitAll(Unknown Source)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt$emitAll$1.invokeSuspend(Channels.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:740)
```
Of course this test is not my real code, but this appears to reproduce the exception (almost) every time it is run. You may need to run the test again if you do not get the exception. It should happen within a few seconds of running the test.

Is this expected behaviour and/or is there anything I can do to fix this?

There was a discussion about this on Slack with @qwwdfsad: https://kotlinlang.slack.com/archives/C1CFAFJSK/p1563388954310700

--- a/kotlinx-coroutines-core/common/src/selects/Select.kt
+++ b/kotlinx-coroutines-core/common/src/selects/Select.kt
@@ -311,10 +311,17 @@ internal class SelectBuilderImpl<in R>(
     internal fun handleBuilderException(e: Throwable) {
         if (trySelect(null)) {
             resumeWithException(e)
-        } else {
-            // Cannot handle this exception -- builder was already resumed with a different exception,
-            // so treat it as "unhandled exception"
-            handleCoroutineException(context, e)
+        } else if (e !is CancellationException) {
+            /*
+             * Cannot handle this exception -- builder was already resumed with a different exception,
+             * so treat it as "unhandled exception". But only if  it is not the completion reason
+             *  and it's not the cancellation. Otherwise, in the face of structured concurrency
+             * the same exception will be reported to theglobal exception handler.
+             */
+            val result = getResult()
+            if (result !is CompletedExceptionally || unwrap(result.cause) !== unwrap(e)) {
+                handleCoroutineException(context, e)
+            }
         }
     }
 

--- a/kotlinx-coroutines-core/jvm/test/flow/CombineStressTest.kt
+++ b/kotlinx-coroutines-core/jvm/test/flow/CombineStressTest.kt
@@ -1,0 +1,53 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow
+
+import kotlinx.coroutines.*
+import org.junit.*
+
+class CombineStressTest : TestBase() {
+
+    @Test
+    public fun testCancellation() = runTest {
+        withContext(Dispatchers.Default + CoroutineExceptionHandler { _, _ -> expectUnreached() }) {
+            flow {
+                expect(1)
+                repeat(1_000 * stressTestMultiplier) {
+                    emit(it)
+                }
+            }.flatMapLatest {
+                combine(flowOf(it), flowOf(it)) { arr -> arr[0] }
+            }.collect()
+            finish(2)
+            reset()
+        }
+    }
+
+    @Test
+    public fun testFailure() = runTest {
+        val innerIterations = 100 * stressTestMultiplierSqrt
+        val outerIterations = 10 * stressTestMultiplierSqrt
+        withContext(Dispatchers.Default + CoroutineExceptionHandler { _, _ -> expectUnreached() }) {
+            repeat(outerIterations) {
+                try {
+                    flow {
+                        expect(1)
+                        repeat(innerIterations) {
+                            emit(it)
+                        }
+                    }.flatMapLatest {
+                        combine(flowOf(it), flowOf(it)) { arr -> arr[0] }
+                    }.onEach {
+                        if (it >= innerIterations / 2) throw TestException()
+                    }.collect()
+                } catch (e: TestException) {
+                    expect(2)
+                }
+                finish(3)
+                reset()
+            }
+        }
+    }
+}

